File: plump/network.py
def send_to_remote(socket, text):
    data = text.encode("utf-8")
    while len(data) > 0:
        sent = socket.send(data)
        data = data[sent:]


def send(socket, text):
    send_to_remote(socket, text) if socket else print(text, end="")


def readline_from_remote(socket):
    all = b""
    while True:
        received = socket.recv(1024)
        all += received
        if received[-1] == b"\n"[0]:
            return all.decode("utf-8")


def readline(socket):
    return (readline_from_remote(socket) if socket else input()).strip()


def readline_with_prompt(socket, prompt):
    send(socket, prompt)
    return readline(socket)

File: plump-rs/src/network.rs
use std::{
    io::{stdin, stdout, Error as IoError, ErrorKind, Read, Result as IoResult, Write},
    net::TcpStream,
};

pub fn input(prompt: &str) -> IoResult<String> {
    print!("{}", prompt);
    stdout().flush()?;

    let mut result = String::new();
    stdin().read_line(&mut result)?;

    Ok(result)
}

pub fn send_to_remote(socket: &mut TcpStream, text: String) -> IoResult<()> {
    let mut data = text.into_bytes();
    while !data.is_empty() {
        let sent = socket.write(&data)?;
        data.drain(0..sent);
    }

    Ok(())
}

pub fn readline_from_remote(socket: &mut TcpStream) -> IoResult<String> {
    const NEWLINE: u8 = 0xA;
    let mut all = Vec::<u8>::new();
    loop {
        let mut buffer = [0u8; 1024];
        let received = socket.read(&mut buffer)?;
        all.extend_from_slice(&buffer[0..received]);

        if all.last() != Some(&NEWLINE) {
            continue;
        }

        return String::from_utf8(all).map_err(|_| IoError::new(ErrorKind::Other, "invalid UTF-8"));
    }
}
